upstream-sha256: 5f789662185e38e72057c7cf8a4728a64d275668ff32946d0f4158c47603b93d
actual-sha256: d57509d723b10b450a3fab5207b1e232f5417fd70b19c776d9a559f0f40424bf
--- upstream.expected
+++ tolk.actual
@@ -12,7 +12,7 @@
   float val0 = (*(data1_256+(gidx1+alu0)));
   int gidx0 = blockIdx.x; /* 16 */
   float val1 = (*(data2_256+(gidx0+alu0)));
-  *(buf2+lidx0) = (val0*val1);
+  *(buf2+lidx0) = (0.0f+(val0*val1));
   __syncthreads();
   *(buf1+0) = 0.0f;
   for (int Ridx104 = 0; Ridx104 < 16; Ridx104++) {
